zarr/core.py

```diff
@@ -1859,9 +1879,38 @@ def _chunk_setitems(self, lchunk_coords, lchunk_selection, values, fields=None):
         ckeys = [self._chunk_key(co) for co in lchunk_coords]
         cdatas = [self._process_for_setitem(key, sel, val, fields=fields)
                   for key, sel, val in zip(ckeys, lchunk_selection, values)]
-        values = {k: v for k, v in zip(ckeys, cdatas)}
+        values = {}
+        if not self._write_empty_chunks:
+            for ckey, cdata in zip(ckeys, cdatas):
+                if self._chunk_isempty(cdata) and not self._chunk_delitem(ckey):
+                    values[ckey] = self._encode_chunk(cdata)
+        else:
+            values = dict(zip(ckeys, map(self._encode_chunk, cdatas)))
         self.chunk_store.setitems(values)
 
+    def _chunk_isempty(self, chunk):
+        if self.dtype == 'object':
+            # we have to flatten the result of np.equal to handle outputs like
+            # [np.array([True,True]), True, True]
+            is_empty = all(flatten(np.equal(chunk, self.fill_value, dtype='object')))
+        else:
+            is_empty = np.all(chunk == self._fill_value)
```

Comment:
does this work properly if `fill_value` is NaN?

Reply:
Nope! Do you have any advice for making that work?

Reply:
Maybe something like this?

```suggestion
            is_empty = np.all(chunk == self._fill_value if np.isnan(self._fill_value) else np.isnan(chunk))
```